### PR TITLE
Display exit code with InvocationErrors

### PR DESCRIPTION
**Pull request by ederag**
This is a PR for @blueyed patch (Issue #290), supplemented with tests. 
For exit codes larger than 128, a hint is given: on unix systems this often means a fatal error.
(Not always, since an application might exit with a large custom error code)

Not showing this hint for windows platform was considered, but discarded to keep code short and
simplify tests.

Before:

> ERROR: InvocationError: '/home/ederag/share/prog/python/tox/.tox/exitcode/bin/python3.6 -c import sys; sys.exit(5)'

After:

> ERROR: InvocationError for command '/home/ederag/share/prog/python/tox/.tox/exitcode/bin/python3.6 -c import sys; sys.exit(5)' (exited with code 5)

or, for an exit code larger than 128:

> ERROR: InvocationError for command '/home/ederag/share/prog/python/tox/.tox/exitcode/bin/python3.6 -c import sys; sys.exit(129)' (exited with code 129)
> Note: On unix systems, an exit code larger than 128 often means a fatal error (e.g. 139=128+11: segmentation fault)

To generate these outputs, a section was added to `tox.ini`:
```
[testenv:exitcode]
basepython = python3.6
description = commands with several exit codes
skip_install = True
commands = python3.6 -c "import sys; sys.exit(5)"
```
and run with `python3 -m tox -e exitcode`.
The modified section is too specific to be included in the PR, isn't it ?

Tests have been added to `test_z_cmdline.py`.
Safely replacing these tests with faster ones (mock) would require someone experienced with this technique.

The documentation might be completed (still have to think about it).
Probably add a section "Understanding InvocationError exit codes"
in [example/general.html](https://tox.readthedocs.io/en/latest/example/general.html)
with links to `pytest` exit codes, to http://www.faqs.org/docs/abs/HTML/exitcodes.html, 
and explaining that no exit code is shown when the process has been aborted (like Qt does).
And for good measure, giving a link to [ignoring-a-command-exit-code](https://tox.readthedocs.io/en/latest/example/basic.html#ignoring-a-command-exit-code).

## Contribution checklist:

- [x] wrote descriptive pull request text
- [x] added/updated test(s)
- [x] updated/extended the documentation
- [x] added relevant [issue keyword](https://help.github.com/articles/closing-issues-using-keywords/)
      in message body
- [x] added news fragment in [changelog folder](https://github.com/tox-dev/tox/tree/master/changelog)
  * fragment name: `<issue number>.<type>.rst` for example (588.bugfix.rst)
  * `<type>` is must be one of `bugfix`, `feature`, `deprecation`,`breaking`, `doc`, `misc`
  * if pr has no issue: consider creating one first or change it to the pr number after creating the pr
  * "sign" fragment with "by @<your username>"
  * please use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files - by @superuser."
  * also see [examples](https://github.com/tox-dev/tox/tree/master/changelog/examples.rst)
- [x] added yourself to `CONTRIBUTORS` (preserving alphabetical order)

### CODE REVIEW

File: tox/__init__.py [hunk 1]
```diff
@@ -34,6 +34,22 @@ class InterpreterNotFound(Error):
 
     class InvocationError(Error):
         """ an error while invoking a script. """
+        def __init__(self, *args):
+            super(exception.Error, self).__init__(*args)
+            self.command = args[0]
+            try:
+                self.exitcode = args[1]
+            except IndexError:
+                self.exitcode = None
```
Comment (asottile) [inline tox/__init__.py]: I think this would be a little more clear as:

```python
        def __init__(self, command, exitcode=None):
            super(Error, self).__init__(command, exitcode)
            self.command = command
            self.exitcode = exitcode
```

File: tox/__init__.py [hunk 2]
```diff
@@ -34,6 +34,22 @@ class InterpreterNotFound(Error):
...
+
+        def __str__(self):
+            str_ = "%s for command %s" % (self.__class__.__name__, self.command)
+            if self.exitcode:
+                str_ += " (exited with code %d)" % (self.exitcode)
+                if self.exitcode > 128:
+                    str_ += ("\nNote: On unix systems, an exit code larger than 128 "
+                             "often means a fatal error (e.g. 139=128+11: segmentation fault)")
```
Comment (asottile) [inline tox/__init__.py]: >128 means it exited due to a fatal signal

You can see the various signals that might trigger this in the table at the bottom of [this](https://en.wikipedia.org/wiki/Signal_(IPC)#POSIX_signals)

We could also compute this and show it to the user:

```console
$ python -c 'import time; time.sleep(100)'
^\Quit (core dumped)
$ echo $?
131
$ python -c '
> import signal
> sigs = {n: s for s, n in vars(signal).items() if s.startswith("SIG")}
> print(sigs[131 - 128])
> '
SIGQUIT
```

A `KeyError` on `sigs` (in my example) means it's probably not a fatal signal that killed it (also note that although `^C` sends SIGINT -- python sets a signal handler for this and reraises it as `KeyboardInterrupt` leading to an exit code of `1` instead of the expected `130`)

File: tox/__init__.py (hunk 1 above)
Comment (ederag) [reply tox/__init__.py]: Thanks a lot. Just had to use `exception.Error`.

Comment (asottile) [reply tox/__init__.py]: oops yeah, sorry my sample wasn't completely copy-pastable 😆

File: tox/__init__.py (hunk 2 above)
Comment (ederag) [reply tox/__init__.py]: Thanks for the link, added to the documentation that was in preparation.
Giving the signal value/name could be nice, since python uses directly the relevant `signals.h`; got to think about it.

**Comment by asottile** [comment]
@gaborbernat I had an outstanding issue, I don't think this was ready for merge

**Comment by gaborbernat** [comment]
@asottile which one?

**Comment by asottile** [comment]
I suggested an alternate wording and some additional helpful output [here](https://github.com/tox-dev/tox/pull/760#discussion_r165831005)

**Comment by ederag** [comment]
TL;DR: done, works, but could be better.
The [commit](https://github.com/ederag/tox/commit/bf9202ac267d461d9e8890e93605023b1ad95a4f) does not appear. Is this PR broken ?

About the proposed wording "means it exited due to a fatal signal", it is true that "signal" is more precise than "error". Finally, the wording "fatal error signal" of [faqs.org](http://www.faqs.org/docs/abs/HTML/exitcodes.html), seems good (especially for segmentation fault, since it originates from an error in the program, in the first place).

The note now gives the corresponding signal if one is found, hopefully matching your suggestion.
Issuing that message only for unix would require mocking `sys.platform` in tests.
Note sure it is worth the trouble though. And it would require [`pytest.mock`](https://pypi.python.org/pypi/pytest-mock).

Now testing the `InvocationError` itself is done more extensively, and much faster, in `test_result.py`. Obviously, it is not a `tox.result`, but kind of a result. Moreover, no other file (except the slow `test_z_cmdline.py`) seemed adequate.
That `InvocationError ` is actually called or not, is still tested in `test_z_cmdline.py`. These tests look for a specific string in the tox output.
Two tests - exit code 0 or non-zero - remain. To be faster, it would be possible to use both commands in the same test. But it seemed safer to keep these two separate.
One fast and safe way would have been to use `mock`, but it did not work because [exceptions can not be mocked like that](https://stackoverflow.com/a/31873937/3565696). One way to circumvent that would be to create a separate *function* `tox.exceptions.InvocationError._str(command, exitcode)`, called from `tox.exceptions.InvocationError.__str__`, and use the [`pytest.mock`](https://pypi.python.org/pypi/pytest-mock) `spy` on the `_str` function only.
This would allow to gain about 3 second of test time for each python version tested.

Questions:
Just in case, would it be acceptable to add [`pytest.mock`](https://pypi.python.org/pypi/pytest-mock) to the `extra_require['testing']` in `setup.py` ?

Finally, a minor question from the initial post is still pending: 

> The modified section is too specific to be included in the PR, isn't it ?

**Comment by gaborbernat** [comment]
@ederag  I think it's because the PR is closed, please open a new one :+1:

**Comment by ederag** [comment]
Waiting for revert-760-exitcode to be merged into master. Glad I checked before pulling. This is a bit weird. Shouldn't it be automatic when the revert button is used ? Why can't we keep the current PR ? A github limitation or bug ?

**Comment by gaborbernat** [comment]
no need to have the merge, just create a PR against the master.

**Comment by ederag** [comment]
Strange. If you won't merge `revert-760-exitcode`, then why revert the PR in the first place ?

**Comment by gaborbernat** [comment]
Was a miss click creation of that branch,  sorry if it caused confusion, the revert was never merged on master https://github.com/tox-dev/tox/commits/master, I consider the extra changes here an improvement over what we have here. Hence  why saying just  create a new PR 👍

**Comment by ederag** [comment]
OK, then the simplest way seems to make a brand new fork and paste the modified files in it. Will file a new PR later.

File: tox/__init__.py (hunk 2 above)
Comment (blueyed) [reply tox/__init__.py]: For reference: subprocess actually returns negative exit codes for signals.
See https://github.com/tox-dev/tox/pull/1401 for a followup/fix.
